### PR TITLE
Provide better parentPath typings

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

In this PR we provide better `NodePath#parentPath` and `NodePath#parent` typings using a map from node types to its parent AST node types, generated from our AST definitions. There are no runtime overhead shipped to our end users because they are done in typings.

The new typings catches two babel types error because we used the `parentPath` info in certain plugins but forget to add babel types support.

This PR includes commits from #14737. I will rebase once that PR is merged.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14739"></a>

### CODE REVIEW

File: packages/babel-types/scripts/generators/ast-types.js [hunk 1]
```diff
@@ -123,6 +158,21 @@ export interface ${deprecatedAlias[type]} extends BaseNode {
     t.DEPRECATED_KEYS
   ).join(" | ")}\n\n`;
 
+  code += "export interface ParentMaps {\n";
+
+  registerParentMaps("null", [...Object.keys(t.DEPRECATED_KEYS)]);
+  // todo: provide a better parent type for Placeholder, currently it acts
+  // as a catch-all parent type for an abstract NodePath, s.t NodePath.parent must
+  // be a Node if type has not been specified
+  registerParentMaps("Node", ["Placeholder"]);
```
Comment (JLHwung) [inline packages/babel-types/scripts/generators/ast-types.js:167]: Caveat: the Placeholder has a unique role here.

File: packages/babel-plugin-transform-typescript/src/enum.ts [hunk 2]
```diff
@@ -156,7 +156,10 @@ export function translateEnumValues(
       } else {
         const initializerPath = memberPath.get("initializer");
 
-        if (initializerPath.isReferencedIdentifier()) {
+        if (
+          initializerPath.isIdentifier() &&
+          initializerPath.isReferencedIdentifier()
+        ) {
           ReferencedIdentifier(initializerPath, {
```
Comment (JLHwung) [inline packages/babel-plugin-transform-typescript/src/enum.ts:163]: `isReferencedIdentifier` ensures that `this` is either an `Identifier` or a `JSXIdentifier`, but `ReferencedIdentifier` can't handle `JSXIdentifier`.

File: packages/babel-types/src/ast-types/generated/index.ts [hunk 3]
```diff
@@ -2780,3 +2781,5773 @@ export type DeprecatedAliases =
...
+    | TypeParameter
+    | TypeParameterInstantiation
+    | TypeofTypeAnnotation
+    | UnionTypeAnnotation;
+  Super:
+    | ArrayExpression
+    | ArrowFunctionExpression
+    | AssignmentExpression
+    | AssignmentPattern
+    | AwaitExpression
```
Comment (JLHwung) [inline packages/babel-types/src/ast-types/generated/index.ts:6437]: This is incorrect. A `Super`'s parent must be either a CallExpression or a MemberExpression. We should consider remove `Super` from `Expression` in Babel 8.

File: packages/babel-plugin-transform-typescript/src/enum.ts (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-typescript/src/enum.ts]: I wonder why our `NodePath` has the wrong variance and `NodePath<A | B>` is not inferred as the same as `NodePath<A> | NodePath<B>` :thinking:

**Review by nicolo-ribaudo — DISMISSED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by liuxingbaoyu — APPROVED** [review]

**Pull request merged**